Codec and Response fixes

Konicai opened the pull request:
Custom Form:
- `deepCopy()` (used in the codec deserializer) as only been public since gson 2.8.2, which isn't present on older server versions (I think 1.17.1 and lower but it doesn't matter too much). The copied array is only used for counting by removing elements so this was avoided pretty easily.
- `valueAt` in the response used to return `AbsentComponent` instead of null. That would then make any of the methods for getting a response, that accept an index, throw a wrong type instead of returning a default, contradicting their javadoc. e.g:

```java
Form form = CustomForm.builder()
    .title("Test Form")
    .label("This form should be empty")
    .optionalInput("Enter something", "placeholder", "default text", false)
    .validResultHandler(response -> {
        player.sendMessage("You entered: " + response.asInput()); // ignores label
        player.sendMessage("You entered: " + response.asInput(1));
    })
    .build();
```
<details>
  <summary>would result in this</summary>
  
```[19:33:33 ERROR]: [floodgate] Error while processing form response!
java.lang.IllegalStateException: Expected input on 1, got org.geysermc.cumulus.util.AbsentComponent@2ddce814
        at org.geysermc.cumulus.response.impl.CustomFormResponseImpl.wrongType(CustomFormResponseImpl.java:329) ~[?:?]
        at org.geysermc.cumulus.response.impl.CustomFormResponseImpl.asInput(CustomFormResponseImpl.java:266) ~[?:?]
        at dev.kejona.crossplatforms.command.defaults.TestCommand.lambda$register$1(TestCommand.java:46) ~[?:?]
        at org.geysermc.cumulus.form.impl.FormImpl$Builder.lambda$setResponseHandler$1(FormImpl.java:273) ~[?:?]
        at org.geysermc.cumulus.form.impl.FormImpl.callResultHandler(FormImpl.java:68) ~[?:?]
        at org.geysermc.cumulus.form.impl.FormDefinition.callResponseHandler(FormDefinition.java:74) ~[?:?]
        at org.geysermc.cumulus.form.impl.FormDefinition.handleFormResponse(FormDefinition.java:62) ~[?:?]
        at org.geysermc.floodgate.pluginmessage.channel.FormChannel.callResponseConsumer(FormChannel.java:130) ~[?:?]
        at org.geysermc.floodgate.pluginmessage.channel.FormChannel.handleServerCall(FormChannel.java:93) ~[?:?]
        at org.geysermc.floodgate.pluginmessage.SpigotPluginMessageRegistration.lambda$register$0(SpigotPluginMessageRegistration.java:44) ~[?:?]
        at org.bukkit.plugin.messaging.StandardMessenger.dispatchIncomingMessage(StandardMessenger.java:436) ~[patched_1.13.2.jar:git-Paper-657]
        at net.minecraft.server.v1_13_R2.PlayerConnection.a(PlayerConnection.java:2625) ~[patched_1.13.2.jar:git-Paper-657]
        at net.minecraft.server.v1_13_R2.PacketPlayInCustomPayload.a(SourceFile:45) ~[patched_1.13.2.jar:git-Paper-657]
        at net.minecraft.server.v1_13_R2.PacketPlayInCustomPayload.a(SourceFile:9) ~[patched_1.13.2.jar:git-Paper-657]
        at net.minecraft.server.v1_13_R2.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:15) ~[patched_1.13.2.jar:git-Paper-657]
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:1.8.0_331]
        at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:1.8.0_331]
        at net.minecraft.server.v1_13_R2.SystemUtils.a(SystemUtils.java:108) ~[patched_1.13.2.jar:git-Paper-657]
        at net.minecraft.server.v1_13_R2.MinecraftServer.b(MinecraftServer.java:1018) ~[patched_1.13.2.jar:git-Paper-657]
        at net.minecraft.server.v1_13_R2.DedicatedServer.b(DedicatedServer.java:439) ~[patched_1.13.2.jar:git-Paper-657]        at net.minecraft.server.v1_13_R2.MinecraftServer.a(MinecraftServer.java:940) ~[patched_1.13.2.jar:git-Paper-657]        at net.minecraft.server.v1_13_R2.MinecraftServer.run(MinecraftServer.java:837) ~[patched_1.13.2.jar:git-Paper-657]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_331]
```
  
</details>

I think that CustomForm Codec/Response still has some issues imo.
- It seems null is not inserted into `componentTypes` for optional components that were not added. Afaik this means it won't actually line up with the contents of `responses` if an optional component is added to the form. 

Less problematic things that shouldn't be encountered if you're doing things right:
- ~~The usage of null by both label and absent components also means that methods for getting responses will return a default instead of throwing a wrong type exception when used on the index of a label component. This isn't as problematic though.~~ Fixed.
- Also, both `asDropdown` and `asStepSlider` can be used on both dropdowns and step sliders without throwing a wrong type exception, which contradicts their javadocs. If `componentTypes` was fixed then this could be fixed.
---
   
Simple Form:
- The deserializer would continue iterating over the button list when the correct button id had already been found.

---
  
Fixes https://github.com/GeyserMC/Floodgate/issues/322

Konicai committed:
Avoid JsonArray deepCopy()

Konicai committed:
Fix CustomFormResponse#valueAt for absent components

Camotoy commented:
Have you tested with... 1.8.8?

Konicai commented:
Yes

Tim203 commented:
Can you change the pom to use the Gson version used by Spigot 1.8? This way something like this couldn't happen again.
The `componentTypes` method is deprecated since 1.1 and isn't supposed to be used for new projects, so it isn't updated for the new features.
And to fix the label & absent returning null thing: we could make an internal `valueAt` which works just like it works atm in master, and a public `valueAt` that works like it does in this PR.
As for `asDropdown` and `asStepSlider`: we could update `componentTypes` in 2.0 (once the public method is removed) but I thought that it wasn't a big deal.

Konicai commented:
> Can you change the pom to use the Gson version used by Spigot 1.8? This way something like this couldn't happen again.
> The `componentTypes` method is deprecated since 1.1 and isn't supposed to be used for new projects, so it isn't updated for the new features.
> And to fix the label & absent returning null thing: we could make an internal `valueAt` which works just like it works atm in master, and a public `valueAt` that works like it does in this PR.

Sounds good

> As for `asDropdown` and `asStepSlider`: we could update `componentTypes` in 2.0 (once the public method is removed) but I thought that it wasn't a big deal.

Yeah it isn't really

Konicai committed:
Drop gson to 2.3.1 (Provided by Spigot 1.8.8)

Konicai committed:
Throw when using `as` response getter on Labels, CustomFormResponse tests

Also allow CustomFormResponse#index to be provided an int value of -1, which is the default when constructed.

Konicai commented:
- gson has been dropped to 2.3.1

CustomFormResponse:
- `nextOrAbsent` and `valueOrAbsent` have been introduced internally so that `AbsentComponent` is provided to `asInput`, `asDropdown` (etc) when applicable
- `asInput`, `asDropdown` (etc) now throw when used on a label component, and return their default when used on `AbsentComponent`
- `index` now allows values of -1 and above, since -1 is the default. This way you can reset the index and use `next()` to get the *first* component.
- Tests to assert wrong type is always thrown when it should be, and assert correct and default values (where applicable)

Konicai committed:
Remove outdated comment

Konicai committed:
Don't call get twice when correct button is found

Tim203 commented on src/main/java/org/geysermc/cumulus/response/impl/CustomFormResponseImpl.java:
```diff
@@ -160,11 +184,10 @@ public <T> T next() {
       return 0;
-    }
-    if (next instanceof Integer) {
+    } else if (next instanceof Integer) {
```
Can you remove the `else` from the `else if`?
The else is not needed because the `if` always returns.

(Please apply this comment for the whole class)

Tim203 commented on src/test/java/org/geysermc/cumulus/response/CustomFormResponseTest.java:
```diff
@@ -0,0 +1,216 @@
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.function.Executable;
+
+public class CustomFormResponseTest {
```
I like the idea behind this, but I'm not convinced by the way this currently looks.
I think the best idea for now is to exclude the test part and we'll look at testing another time, so that we can get this merged soon.

Tim203 commented on src/main/java/org/geysermc/cumulus/response/impl/CustomFormResponseImpl.java:
```diff
@@ -326,9 +353,17 @@ public boolean getToggle(@NonNegative int index) {
+    Object response = responses.get(index);
+    String unexpected;
+    if (response == null) {
+      unexpected = "a Label";
```
Change this to `label` to stay consistent with the other calls of this method.

Konicai committed:
Remove unnecessary `else if`, update wrongType for labels, remove tests
